# System Recommenders - Final Project 2025

### 🎯 Objective

Develop a recommender system that suggests short videos to users based on user preferences, interaction histories, and video content using the KuaiRec dataset. 

The challenge is to create a personalised and scalable recommendation engine similar to those used in platforms like TikTok or Kuaishou.

### 📥 Imports

In [5]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
from scipy.sparse import csr_matrix
from implicit.als import AlternatingLeastSquares

plt.rcParams["figure.figsize"] = (20, 13)
%matplotlib inline
%config InlineBackend.figure_format = "retina"

### 📊 Download Dataset

We will use the **KuaiRec dataset**, a large-scale, fully-observed dataset collected from the Kuaishou short-video platform.

It contains:

- **User interactions** (views, likes, etc.)
- **Video metadata** (video ID, tags, etc.)
- **Timestamps**

More info: [KuaiRec Paper](https://arxiv.org/abs/2202.10842)

**Download dataset**

1. <ins>First option : Downloading Dataset via wget<ins>

In [6]:
%%bash
if [ ! -d "./data_final_project" ]; then
  wget --no-check-certificate 'https://drive.usercontent.google.com/download?id=1qe5hOSBxzIuxBb1G_Ih5X-O65QElollE&export=download&confirm=t&uuid=b2002093-cc6e-4bd5-be47-9603f0b33470' -O KuaiRec.zip
  unzip KuaiRec.zip -d ../data_final_project
else
  echo "Directory './data_final_project' already exists. Skipping download."
fi

Directory './data_final_project' already exists. Skipping download.


2. <ins>Second option : Downloading dataset via Google Drive<ins>

If the data is not downloaded by the wget because of a Connection Refused you might download it via this [link](https://drive.google.com/file/d/1qe5hOSBxzIuxBb1G_Ih5X-O65QElollE/view) and place it at the project root.

In [7]:
""" Uncomment if you need to (and if the first option did not work correctly)
%%bash
unzip KuaiRec.zip -d data_final_project
mv "data_final_project/KuaiRec 2.0/" data_final_project/KuaiRec
"""

' Uncomment if you need to (and if the first option did not work correctly)\n%%bash\nunzip KuaiRec.zip -d data_final_project\nmv "data_final_project/KuaiRec 2.0/" data_final_project/KuaiRec\n'

From this dataset we obtain the following files :

```bash
KuaiRec
  ├── data
  │   ├── big_matrix.csv          
  │   ├── small_matrix.csv
  │   ├── social_network.csv
  │   ├── user_features.csv
  │   ├── item_daily_features.csv
  │   └── item_categories.csv
  │   └── kuairec_caption_category.csv
```

- `interactions_train.csv`: historical user-item interactions for training.
- `interactions_test.csv`: user-item pairs to score during testing.
- `sample_submission.csv`: a template showing the expected output format.
- `video_metadata.csv`: metadata including tags or content-related features.

![image](img/KuaiRec.png)

## **1️⃣ Dataset Preprocessing**
📝 Associated Tasks :
- Load and inspect the dataset.
- Handle missing or inconsistent data.
- Merge metadata for content-based models if necessary.

---


<div style="background-color:#cccccc; padding: 10px; border-radius: 5px; border: 1px solid #aaa; font-weight: bold; color:#111;">
    💡 For more details about the different analyses and pre-processing decisions made on the available datasets, <a href="../EDA/EDA.ipynb" style="color:#1a73e8;">feel free to click here and go to the EDA notebook</a>.
</div>


---

However, here is a small recap of the observations made : 

- **User Interactions**: 
  - The majority of users have **fewer than 3,000 interactions**.
  - There are a few **outliers** with interactions exceeding 6,000.

- **Item Popularity**:
  - There are very few items that receive **high levels of interaction**.
  - Many items, while not extremely popular, still receive a fair amount of attention.
  - A small subset of items are truly **highly popular**.

- **Time-Based Trends**:
  - However, interactions peak towards the **end of the week**, specifically from **Friday to Sunday**.
  - **Late-night hours (0:00 - 3:00 a.m.)** see the highest levels of activity.
  - **From 10:00 to 20:00** we see less activity
  - The time of day or day of the week has **minimal to no impact** on the watch ratio.

- **User Behavior**:
  - The **top 10 most active users** show a distinct dip in activity around **12:00 p.m.** and **4:00 p.m.**.

- **Video Length**:
  - **Shorter videos** (up to 30 seconds) receive **more interactions** and have a **higher watch ratio**.
  - Conversely, **longer videos** tend to have **fewer interactions** and a lower watch ratio.
  - Videos that are **no longer than 30 seconds** appear to have the optimal watch ratio.

- **Watch Ratio**:
  - **Half of the videos have less then 75% watch ratio.**
  - Only **0.1% of the videos have more than 18** of watch ratio. These are extreme outliers, we can either drop them or try and normalize them.

- **Video Type:**

    - **"Ad" videos have significantly lower interaction** rates compared to Normal videos, which are more engaging for users.
    - Users tend to engage less with promotional or advertisement-type content.

- **User Preferences:**

    - There is a clear preference for **Short Imports videos**. These videos are viewed and interacted with more often compared to longer or different upload types.

- **Video Format:**

    - The **1280x720 resolution is the optimal format** for maximizing user interaction and watch ratio. Other formats either fall short or show no significant improvement.

- **Video Privacy Settings:**

    - As expected, **public videos receive more interactions** than private or only friends videos. This could be due to the higher volume of public content being uploaded, which in turn increases the overall interactions.

    - Private or only friends videos tend to have a smaller, more targeted audience, reducing their interaction rates.

- **Video Age**:
    - **Recent videos (lower video age in days) tend to have more interactions**



### Load Datasets

Here as we have shown in the [EDA notebook](./EDA/EDA.ipynb) with more details :

1. We have to load the dataset we will use.

    We will only be using three datasets :
    - `big_matrix.csv`
        - Corresponds to the different interactions made on the videos
        - It will serve for the training
    - `small_matrix.csv`
        - Same as big_matrix
        - It will serve for the testing
    - `item_daily_features.csv`
        - Contains informations about the video (e.g. number of likes, shares, reports ...) 
        - It will be merged both with the small and big matrix to give additional informations

2. We make small preprocessing :
    - Converting some columns to a more appropriate type
    - Removing highly correlated columns
    - We will cap videos with an above 2.34 watch ratio as above are only the top 5% as show in the [EDA notebook](./EDA/EDA.ipynb). This will help normalize the `watch_ratio` which is an important metric.


In [8]:
interactions = pd.read_csv("./data_final_project/KuaiRec/data/big_matrix.csv")
small_interactions = pd.read_csv("./data_final_project/KuaiRec/data/small_matrix.csv")
item_features = pd.read_csv("./data_final_project/KuaiRec/data/item_daily_features.csv")

def clean_df(df):
    df = df.dropna()
    df = df.drop_duplicates()  
    return df  

def clean_df_timestamp(df):
    df = clean_df(df)
    df = df[df["timestamp"] >= 0]
    df = df[df["watch_ratio"] <= 200]
    return df

item_features = clean_df(item_features)
item_features = item_features.drop_duplicates(subset='video_id')
train_df = clean_df_timestamp(interactions)
test_df = clean_df_timestamp(small_interactions)

item_features['upload_dt'] = pd.to_datetime(item_features['upload_dt'])
item_features['date'] = pd.to_datetime(item_features['date'], format='%Y%m%d')


We drop some columns in the test and train dataframes that we will not need.

In [9]:
to_drop = ['date', 'play_duration', 'video_duration', 'time', 'timestamp']

train_df.drop(columns=to_drop, inplace=True, errors='ignore')
test_df.drop(columns=to_drop, inplace=True, errors='ignore')

We apply a cap to 2.34 to the watch ratio, to prevent outliers.

This decision is mostly made to not only suggest videos that have a way too high watch ratio 
(some of the videos have a 600 watch ratio) but also some videos that have a smaller watch ratio.

In [10]:
train_df['watch_ratio'] = train_df['watch_ratio'].apply(lambda x: min(x, 2.34))
test_df['watch_ratio'] = test_df['watch_ratio'].apply(lambda x: min(x, 2.34))

We drop columns that have high correlation between them.

In [11]:
correlation = item_features[[
       'video_duration', 'video_width',
       'video_height', 'music_id',
       'show_cnt', 'show_user_num', 'play_cnt', 'play_user_num',
       'play_duration', 'complete_play_cnt', 'complete_play_user_num',
       'valid_play_cnt', 'valid_play_user_num', 'long_time_play_cnt',
       'long_time_play_user_num', 'short_time_play_cnt',
       'short_time_play_user_num', 'play_progress', 'comment_stay_duration',
       'like_cnt', 'like_user_num', 'click_like_cnt', 'double_click_cnt',
       'cancel_like_cnt', 'cancel_like_user_num', 'comment_cnt',
       'comment_user_num', 'direct_comment_cnt', 'reply_comment_cnt',
       'delete_comment_cnt', 'delete_comment_user_num', 'comment_like_cnt',
       'comment_like_user_num', 'follow_cnt', 'follow_user_num',
       'cancel_follow_cnt', 'cancel_follow_user_num', 'share_cnt',
       'share_user_num', 'download_cnt', 'download_user_num', 'report_cnt',
       'report_user_num', 'reduce_similar_cnt', 'reduce_similar_user_num',
       'collect_cnt', 'collect_user_num', 'cancel_collect_cnt',
       'cancel_collect_user_num']].corr()

upper = correlation.where(np.triu(np.ones(correlation.shape), k=1).astype(bool))

to_drop = [column for column in upper.columns if any(upper[column] > 0.8)]
# These columns are dropped as we don't need them anymore 
# (merged in previous cell or just not needed for collaborative-filtering)
item_features.drop(columns=to_drop, inplace=True, errors='ignore')


## **2️⃣ Feature Engineering**
📝 Associated Tasks :
- Create meaningful features from interaction and metadata (e.g., content tags, user activity history).
- Build user-item interaction matrix.
- Optionally extract time-based or popularity-based features.

---

### Basic additional features

As suggested, we will now create some new features for the datasets, the first two will correspond respectively to the `video_age` and a boolean `is_short_video` to see if the video is shorter then 30 seconds because as we saw in the [EDA notebook](./EDA/EDA.ipynb), shorter and recent videos are more interacted with then other type of videos.

In [12]:
item_features['video_age'] = (item_features['date'] - item_features['upload_dt']).dt.days
item_features['is_short_video'] = (item_features['video_duration'].fillna(0) <= 30).astype(int)

We remove those columns as we don't need them in the future.

In [13]:
to_drop =   [
                'date', 'upload_dt', 'video_duration', 'music_id',
                'video_tag_name', 'play_progress', 'video_tag_id',
                'time', 'play_duration'
            ]
item_features.drop(columns=to_drop, inplace=True, errors='ignore')

Let's now merge the item_features dataset, containing some insights on the videos such as number of likes, to our train_df and test_df.

In [14]:
train_df = pd.merge(train_df, item_features, on='video_id', how='left')
test_df = pd.merge(test_df, item_features, on='video_id', how='left')

### Engagement Score

To effectively train our ALS (Alternating Least Squares) recommendation model, we need a **proxy rating** to represent user interest or satisfaction with a video. 

We thus can define a **custom engagement score** that combines multiple implicit feedback signals. 

Based on insights from our exploratory data analysis ([see EDA notebook](./EDA/EDA.ipynb)), we define the **engagement score** using a combination of behavioral, content-related, and video metadata features.

These features capture how likely a user is to have positively interacted with a video, making this score a suitable proxy for collaborative filtering.

#### Key features used to compute the engagement score:

- **Watch Ratio**: Proportional score based on the ratio of the video watched. A higher watch ratio directly increases the score.
- **Short Videos (`is_short_video`)**: Users show a preference for short videos, so these receive a small bonus.
- **Video Age**: Newer videos are generally more engaging; the score includes a time decay bonus favoring recent uploads.
- **Video Type**: Videos marked as **"AD"** receive a penalty.
- **Upload Type**: Some formats are known to drive higher engagement (e.g., `"ShortImport"`, `"StartCamera"`), and are weighted accordingly.
- **Visibility Status**: Public videos are more likely to be seen and engaged with, hence receive a bonus.
- **Video Resolution**: The preferred format is `1280x720`. Videos with this resolution or higher receive more interactions.
- **User Interaction Metrics**:
  - Positive interactions (`like_cnt`, `comment_cnt`, `share_cnt`, etc.) are weighted logarithmically to contribute to the score.
  - Negative feedback (`cancel_like_cnt`, `cancel_follow_cnt`, `report_cnt`) reduces the score, also using logarithmic scaling.

Finally, the score is **min-max normalized** between `0` and `1` to ensure comparability across the dataset and improve model convergence.

This engagement score becomes our **implicit rating** that the ALS model will learn to predict.


In [ ]:
def build_engagement_score(df):
    df["engagement_score"] = 0
    
    # Watch Ratio
    if 'watch_ratio' in df.columns:
        df["engagement_score"] += df['watch_ratio'].fillna(0) * 10
    
    # is_short_video
    df["engagement_score"] += df['is_short_video'].fillna(0) * 3
    
    # Video age
    if 'video_age' in df.columns:
        max_age = 365
        normalized_age = np.minimum(df['video_age'].fillna(max_age), max_age) / max_age
        # Newer videos get up to 2 points bonus
        df["engagement_score"] += (1 - normalized_age) * 2
    
    
    if 'video_type' in df.columns:
        df["engagement_score"] += np.where(
            df['video_type'] == 'AD',
            -3,  # penalty for ads
            2    # bonus for regular content
        )
    
    if 'visible_status' in df.columns:
        df["engagement_score"] += np.where(
            df['visible_status'] == 'public',
            2,  
            -1
        )
    
    if 'upload_type' in df.columns:
        upload_type_weights = {
            'ShortImport': 3,     # Short imported videos tend to be high quality
            'StartCamera': 2.5,
            'Knowle': 2,
            'Web': 1.5,
            'LongImport': 1,
            'UNKNOWN': 0,
            'LongCamera': 0.5,
            'PictureSet': 0.5,
            'LongPicture': 0.5,
            'ACurlVideo': 0.5,
            'followShot': 0.5,
            'ShareFromOtherApp': 0.5,
            'SameFrame': 0,
            'PictureCopy': 0,
            'FlashPhoto': 0,
            'PhotoCopy': 0,
            'LocalCollection': 0,
            'LocalInteraction': 0
        }
        df["engagement_score"] += df['upload_type'].map(upload_type_weights).fillna(0)
    
    engagement_columns = {
        'like_cnt': 0.5,
        'comment_cnt': 0.7,
        'share_cnt': 0.8,
        'collect_cnt': 0.6,
        'follow_cnt': 0.9,
        'complete_play_cnt': 0.7,
        'valid_play_cnt': 0.5,
        'reply_comment_cnt': 0.6,
        'comment_like_cnt': 0.4
    }
    
    for col, weight in engagement_columns.items():
        if col in df.columns:
            df["engagement_score"] += np.minimum(np.log1p(df[col].fillna(0)) * weight, 10)


    penalty_columns = {
        'cancel_like_cnt': 0.4,
        'cancel_follow_cnt': 0.5,
        'report_cnt': 0.7
    }
    for col, weight in penalty_columns.items():
        if col in df.columns:
            df["engagement_score"] -= np.minimum(np.log1p(df[col].fillna(0)) * weight, 10)

    if 'video_width' in df.columns:
        df["engagement_score"] += np.where(df['video_width'] >= 720, 0.5, 0)

    if 'video_height' in df.columns:
        df["engagement_score"] += np.where(df['video_height'] >= 1280, 0.5, 0)
    
    # Here we normalize the score    
    min_score = df["engagement_score"].min()
    max_score = df["engagement_score"].max()

    df["engagement_score"] = (df["engagement_score"] - min_score) / (max_score - min_score)
    
    return df

test_df = build_engagement_score(test_df)
train_df = build_engagement_score(train_df)

## **3️⃣ Model Development**
📝 Associated Tasks :
- Choose a recommendation approach:
    - Collaborative filtering (e.g., ALS, Matrix Factorisation)
    - Content-based filtering
    - Sequence-aware models
    - Hybrid approaches
- Train and validate your model on the training set.

---

### ALS Model

The `engagement_score` we made in the previous section will act as the rating of our video. 

Therefore we can use ALS, which is a collaborative filtering model, that will try to guess the engagement score. The higher it is, the more engaging the video is.

The ALS Model was choosen here as it will take the different studied metrics in the one column `engagement_score` which will make more impact.

#### User item matrix 
First we will create the user-item matrix, which is essential for the ALS algorithm, to map the different users, videos and their engagement score.

In [16]:
# Get Unique user and videos
user_ids_train = train_df['user_id'].unique()
video_ids_train = train_df['video_id'].unique()

# Compute index for each user and videos
user_to_index = {user_id: idx for idx, user_id in enumerate(user_ids_train)}
video_to_index = {video_id: idx for idx, video_id in enumerate(video_ids_train)}

# add the index to the train and test
train_df['user_index'] = train_df['user_id'].map(user_to_index)
train_df['video_index'] = train_df['video_id'].map(video_to_index)

test_df['user_index'] = test_df['user_id'].map(user_to_index)
test_df['video_index'] = test_df['video_id'].map(video_to_index)

row = train_df['user_index'].values
col = train_df['video_index'].values

data = train_df['engagement_score'].values

n_users = train_df['user_index'].max() + 1
n_items = train_df['video_index'].max() + 1
    
user_item_matrix = csr_matrix((data, (row, col)), shape=(n_users, n_items))

#### Model Training
Here we train our ALS Model with the `user_item_matrix` 💪

The different parameters were found with the code you can find at the end of this notebook in the  "6️⃣ Trying to improve our model" section.

In [ ]:
model = AlternatingLeastSquares(
    factors=100,
    regularization=0.1,
    iterations=15,
    use_gpu=False,
    alpha=40
)

model.fit(user_item_matrix.T) 

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/cpu/als.py:95: RuntimeWarning: OpenBLAS is configured to use 8 threads. It is highly recommended to disable its internal threadpool by setting the environment variable 'OPENBLAS_NUM_THREADS=1' or by calling 'threadpoolctl.threadpool_limits(1, "blas")'. Having OpenBLAS use a threadpool can lead to severe performance issues here.
  check_blas_config()
/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.05483412742614746 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

## **4️⃣ Recommendation Algorithm**
📝 Associated Tasks :
- Predict which videos are likely to be enjoyed by each user in the test set.
- Generate a top-N ranked list of recommendations for each user.

---
So let's define a simple function to get the top 100 recommendations for each user.

In [ ]:
top_n=100
def get_top_n_recommendations(model, user_item_matrix, user_ids, n=10, seen=True):
    recommendations = {}
    
    for user_id in user_ids:
        # Items the user has already interacted with (adjusted for training)
        already_interacted = set(user_item_matrix[user_id].indices) if seen else set()
        
        # U * V^T => computes the final score
        scores = model.user_factors[user_id].dot(model.item_factors.T)
       
        item_scores = [(item_id, scores[item_id])
                       for item_id in range(len(scores))
                       if item_id not in already_interacted]
        # Sort and select top-N items
        item_scores.sort(key=lambda x: x[1], reverse=True)
        top_items = [item[0] for item in item_scores[:n]]
        
        recommendations[user_id] = top_items

    return recommendations

train_users = train_df['user_index'].unique()
test_users = test_df['user_index'].unique()

# Exclude seen items (for real-world test set evaluation)
test_recommendations = get_top_n_recommendations(
    model, user_item_matrix, test_users, n=top_n, seen=True
)

# Include seen items (to understand model fit on training data)
train_recommendations = get_top_n_recommendations(
    model, user_item_matrix, train_users, n=top_n, seen=False
)

{14: [2267, 594, 122, 6755, 1399, 5767, 1515, 4262, 3591, 4409, 5198, 1829, 2763, 1121, 95, 5726, 4033, 1743, 1097, 7073, 5666, 4216, 6751, 6494, 6844, 2761, 3824, 4854, 6083, 1404, 525, 1131, 4665, 3472, 4917, 1938, 3005, 3730, 3253, 1488, 2830, 2925, 449, 1576, 3819, 1729, 4474, 3716, 3525, 6022, 2908, 4878, 6267, 5175, 5834, 618, 3524, 6785, 729, 6891, 5914, 2896, 5745, 2445, 4696, 2498, 5435, 5470, 2021, 733, 889, 5031, 2417, 1863, 3004, 4947, 170, 6352, 6941, 5748, 3438, 7102, 5392, 887, 5176, 5550, 303, 5064, 6616, 5993, 3743, 3752, 232, 4631, 713, 6697, 6482, 2171, 4314, 6857], 19: [419, 1404, 6308, 2928, 5536, 4181, 4259, 1169, 1455, 1421, 1068, 6726, 590, 5882, 2639, 2856, 6805, 6461, 2406, 958, 6034, 4401, 6996, 5543, 4212, 5411, 6514, 6498, 999, 4029, 3930, 1870, 492, 6300, 1582, 3692, 6289, 2608, 3716, 1214, 6, 5886, 4806, 382, 6577, 3347, 2332, 4376, 31, 5154, 4888, 2471, 2019, 4750, 5354, 3548, 2823, 4199, 5923, 3665, 2914, 593, 4314, 1614, 5645, 6282, 5214, 1204, 690, 63

In [25]:
    
def pretty_print_recommendations(test_recommendations, count=1):    
    for i, (user_id, items) in enumerate(test_recommendations.items()):
        print(f"User {user_id}: {items}")
        if i == count:
            break

pretty_print_recommendations(test_recommendations)    

User 14: [2267, 594, 122, 6755, 1399, 5767, 1515, 4262, 3591, 4409, 5198, 1829, 2763, 1121, 95, 5726, 4033, 1743, 1097, 7073, 5666, 4216, 6751, 6494, 6844, 2761, 3824, 4854, 6083, 1404, 525, 1131, 4665, 3472, 4917, 1938, 3005, 3730, 3253, 1488, 2830, 2925, 449, 1576, 3819, 1729, 4474, 3716, 3525, 6022, 2908, 4878, 6267, 5175, 5834, 618, 3524, 6785, 729, 6891, 5914, 2896, 5745, 2445, 4696, 2498, 5435, 5470, 2021, 733, 889, 5031, 2417, 1863, 3004, 4947, 170, 6352, 6941, 5748, 3438, 7102, 5392, 887, 5176, 5550, 303, 5064, 6616, 5993, 3743, 3752, 232, 4631, 713, 6697, 6482, 2171, 4314, 6857]
User 19: [419, 1404, 6308, 2928, 5536, 4181, 4259, 1169, 1455, 1421, 1068, 6726, 590, 5882, 2639, 2856, 6805, 6461, 2406, 958, 6034, 4401, 6996, 5543, 4212, 5411, 6514, 6498, 999, 4029, 3930, 1870, 492, 6300, 1582, 3692, 6289, 2608, 3716, 1214, 6, 5886, 4806, 382, 6577, 3347, 2332, 4376, 31, 5154, 4888, 2471, 2019, 4750, 5354, 3548, 2823, 4199, 5923, 3665, 2914, 593, 4314, 1614, 5645, 6282, 5214, 1204,

## **5️⃣ Evaluation**
📝 Associated Tasks :
- Choose suitable metrics (e.g., Precision@K, Recall@K, MAP, NDCG).
- Evaluate performance and provide interpretations.

---

For this section we will use several metrics to test our model, these are the metrics and their definitions that can be found in the course or with other ressources :
### 1. **Precision@K**

- Measures the **proportion of relevant items** among the top-K recommended items.

$$
\text{Precision@K} = \frac{\text{number of relevant recommended items in top-K}}{K}
$$

- **High Precision@K** → Users are more likely to find something they like within the first K items.  
- **Best for systems prioritizing “hit rate”** (e.g., e-commerce product suggestions).

--- 

### 2. **Mean Reciprocal Rank (MRR)**  
- Measures the **average rank** of the first relevant item across all queries. It’s especially useful for evaluating systems where the goal is to recommend one relevant item at the top.

$$
\text{MRR} = \frac{1}{|Q|} \sum_{q \in Q} \frac{1}{\text{rank}_q}
$$

Where:  
- Q is the set of queries or users.  
- rank_q is the rank of the first relevant item for query \( q \).

**Why use MRR?**  
- Evaluates how quickly the first relevant item appears. A higher MRR means the system performs better in bringing relevant results to the top.

---

### 3. **Hit Rate**  
- Measures the **proportion of users or queries** that have at least one relevant item in the top-K recommendations.

$$
\text{Hit Rate@K} = \frac{\text{number of queries with at least one relevant item in top-K}}{\text{total number of queries}}
$$

**Why use Hit Rate?**  
- Indicates how frequently the system successfully recommends at least one relevant item. A higher Hit Rate means more users are receiving relevant recommendations in the top-K results.
---

### 4. **Normalised Discounted Cumulative Gain (NDCG@K)**  

- Evaluates the quality of the ranking by giving **higher importance to top-ranked relevant items**.

    $$
    \text{DCG@K} = \sum_{i=1}^{K} \frac{rel_i}{\log_2(i + 1)}
    $$

    $$
    \text{NDCG@K} = \frac{\text{DCG@K}}{\text{IDCG@K}}
    $$

Where:  
- \( rel_i \) = relevance of item at position i (binary or graded relevance)  
- \( IDCG@K \) = DCG of an ideal ranking  

**Why use NDCG?**  
- Penalises relevant items that are **ranked lower**.  
- Captures the fact that users are more likely to click on **top-ranked items**.


In [ ]:
def evaluate_recommendations(recommendations, test_df, top_n=10, k=10):
    # Map of actual items per user
    user_actual_items = test_df.groupby('user_index')['video_index'].apply(set).to_dict()
    
    precision_at_n = []

    # Hit Rate, MRR, nDCG calculations
    hits = 0
    mrr = 0.0
    total_ndcg = 0.0
    count = 0
    
    for user_id, recommended_items in recommendations.items():
        if user_id in user_actual_items:
            actual_items = user_actual_items[user_id]
            recs_at_n = recommended_items[:top_n]

            # Precision
            num_relevant = len(set(recs_at_n) & actual_items)
            precision = num_relevant / len(recs_at_n) if recs_at_n else 0

            precision_at_n.append(precision)
            
            # Hit Rate
            gt = user_actual_items.get(user_id, set())
            if any(item in gt for item in recs_at_n[:k]):
                hits += 1

            # MRR
            for rank, item in enumerate(recs_at_n[:k], start=1):
                if item in gt:
                    mrr += 1.0 / rank
                    break

            # nDCG
            def dcg(recs, gt, k):
                return sum((1 / np.log2(i + 2)) if rec in gt else 0 for i, rec in enumerate(recs[:k]))

            def idcg(gt, k):
                n_relevant = min(len(gt), k)
                return sum(1 / np.log2(i + 2) for i in range(n_relevant))

            idcg_val = idcg(gt, k)
            if idcg_val > 0:
                total_ndcg += dcg(recs_at_n, gt, k) / idcg_val
                count += 1

    avg_precision = np.mean(precision_at_n) if precision_at_n else 0
    hit_rate = hits / len(user_actual_items) if user_actual_items else 0
    avg_mrr = mrr / len(user_actual_items) if user_actual_items else 0
    avg_ndcg = total_ndcg / count if count > 0 else 0

    return avg_precision, hit_rate, avg_mrr, avg_ndcg

print("Evaluating on training set...")
train_precision, train_hit_rate, train_mrr, train_ndcg = evaluate_recommendations(train_recommendations, train_df, top_n=top_n, k=10)
print(f"Training - Precision@{top_n}: {train_precision:.4f}, Hit Rate@{10}: {train_hit_rate:.4f}, MRR@{10}: {train_mrr:.4f}, nDCG@{10}: {train_ndcg:.4f}")


print("Evaluating on test set...")
test_precision, test_hit_rate, test_mrr, test_ndcg = evaluate_recommendations(test_recommendations, test_df, top_n=top_n, k=10)
print(f"Testing - Precision@{top_n}: {test_precision:.4f}, Hit Rate@{10}: {test_hit_rate:.4f}, MRR@{10}: {test_mrr:.4f}, nDCG@{10}: {test_ndcg:.4f}")

Evaluating on training set...
Training - Precision@100: 0.1967, Hit Rate@10: 0.7963, MRR@10: 0.3730, nDCG@10: 0.1992
Evaluating on test set...
Testing - Precision@100: 0.4611, Hit Rate@10: 1.0000, MRR@10: 0.6711, nDCG@10: 0.4662


### Interpretation

We observe that precision over 100 recommendations is lower on the training set than on the testing set, which could be due to matrix sparsity.

The training set, composed of `big_matrix.csv`, is much larger than the testing set, based on `small_matrix.csv`. 

As a result, there are more videos to recommend in the training set, which might impact precision.

- **Precision@100**  
  19% of the videos recommended in the training set and around 46% in the test set are relevant in the top 100 recommendations.

- **Hit Rate@10**  
  79% of users in the training set and 100% in the test set received at least one relevant item in their top 10 recommendations.

- **MRR@10**  
  The mean reciprocal rank for the top 10 recommendations is **0.37** on the training set and **0.67** on the test set.  
  On average, a relevant item appears around the **3rd or 4th** position in the top 10 for the training set, and around the **6th or 7th** position for the test set.

- **nDCG@10**  
  The normalized discounted cumulative gain at rank 10 is **0.1992** in the training set and **0.4662** in the test set.  
  This indicates that the top 10 recommendations are poorly ordered by relevance in the training set, but better ordered in the test set.

---

### Conclusion

- **Precision** is higher for this model on unseen data, suggesting good generalization.
- **Hit Rate** and **MRR** indicate the model consistently presents at least one relevant item in the top recommendations.
- **nDCG** shows that the model has **poor to moderate ranking quality**, especially in the training set.


## 6️⃣ Trying to Improve Our Model

The following code was used to test different parameter configurations of the **implicit ALS** model in order to find the one yielding the best **nDCG** score.

**nDCG** (Normalized Discounted Cumulative Gain) was chosen as the main evaluation metric because it was the most relevant for our context, for several reasons:

- **nDCG considers the position** of relevant items in the recommendation list.

- Since we recommend a short list of top videos to each user, it's important that the most relevant items appear at the top. nDCG is designed to prioritize this type of relevance.

- In real-world applications, users expect the best results early. If they have to scroll too long without finding relevant content, they may quickly lose interest. nDCG reflects this behavior well by penalizing lower-ranked relevant items.

> **In short**: nDCG provides a position-aware evaluation of the quality of our recommendations.

---

As the grid search took around 30 minutes to run, the corresponding code has been commented out for now. 

Moreover this experience was done in the [ALS notebook](models/als.ipynb) so it won't work here.

Below are the best metrics achieved by the model:

> **Best model:**  
> `{'factors': 100, 'regularization': 0.01, 'alpha': 40, 'hit_rate': 0.9978738483345145, 'mrr': 0.6703162791220928, 'ndcg': 0.46914296882680606}`


In [ ]:
"""
best_score = 0
best_params = {}

for factors in [20, 50, 100]:
    for reg in [0.01, 0.1, 0.5]:
        for alpha in [10, 40, 100]:
            print(f"\nTraining ALS with factors={factors}, reg={reg}, alpha={alpha}")
            model = AlternatingLeastSquares(
                factors=factors,
                regularization=reg,
                iterations=15,
                alpha=alpha,
                use_gpu=False
            )
            model.fit(user_item_matrix.T)

            test_recommendations = get_top_n_recommendations(
                model, user_item_matrix, test_users, n=top_n, seen=True
            )

            _, hit_rate, mrr, ndcg = evaluate_recommendations(
                test_recommendations, test_df, top_n=top_n, k=10
            )

            print(f"HitRate@10: {hit_rate:.4f}, MRR@10: {mrr:.4f}, nDCG@10: {ndcg:.4f}")

            if ndcg > best_score:
                best_score = ndcg
                best_params = {
                    'factors': factors,
                    'regularization': reg,
                    'alpha': alpha,
                    'hit_rate': hit_rate,
                    'mrr': mrr,
                    'ndcg': ndcg
                }

print(f"\nBest model: {best_params}")
"""



Training ALS with factors=20, reg=0.01, alpha=10


/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.0623021125793457 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(782, -0.06529804), (5006, -0.056622665), (4435, -0.056085333), (3353, -0.053640276), (2877, -0.053328965), (4898, -0.05258464), (167, -0.052474022), (6602, -0.050295018), (6944, -0.049785074), (3457, -0.0496398), (1231, -0.049157534), (2266, -0.048157725), (4181, -0.048140507), (4259, -0.046282977), (1235, -0.046027258), (740, -0.04548222), (3106, -0.045440182), (5783, -0.04529667), (2099, -0.044802938), (2902, -0.04379806), (2267, -0.043567643), (2128, -0.043005917), (4980, -0.042143233), (897, -0.04184061), (1997, -0.041838486), (6408, -0.041603338), (374, -0.041391682), (826, -0.04138065), (4069, -0.041363645), (4212, -0.04092903), (1356, -0.04084426), (5904, -0.04082241), (3681, -0.040590268), (2062, -0.040425476), (419, -0.04037657), (56, -0.039758198), (227, -0.03966455), (4768, -0.039439797), (1068, -0.039410677), (2915, -0.039250206), (5247, -0.038838256), (6910, -0.03858818), (6828, -0.03846199), (2072, -0.03811736), (1350, -0.038102634), (905, -0.03798085), (1325, -0.037845

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.04934215545654297 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(897, -0.2725075), (6944, -0.24960011), (2381, -0.2327644), (6910, -0.23221502), (1779, -0.23000921), (4898, -0.22925484), (287, -0.21839263), (3457, -0.21621986), (1997, -0.21453875), (2902, -0.21281114), (1949, -0.20269562), (5904, -0.19871445), (5997, -0.19461177), (3353, -0.1931036), (4070, -0.19004391), (980, -0.1854222), (782, -0.17919764), (2536, -0.1789279), (6609, -0.17482147), (2877, -0.17274019), (3845, -0.17246315), (4774, -0.16796763), (2099, -0.16793348), (5156, -0.16565315), (6068, -0.16536483), (6793, -0.16530426), (1325, -0.16215724), (4435, -0.16211212), (4914, -0.16041249), (3681, -0.15992495), (3566, -0.15950161), (5319, -0.15934804), (2745, -0.15828365), (1230, -0.15776016), (2406, -0.15697856), (4089, -0.15632509), (4629, -0.15601373), (6566, -0.15577392), (1571, -0.15297328), (1956, -0.15169159), (5489, -0.15086836), (1733, -0.15068132), (4123, -0.15053031), (6951, -0.1503084), (1231, -0.1485144), (3738, -0.14838347), (1455, -0.14814484), (3135, -0.14762193), (6

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.05083012580871582 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(5904, -0.35387287), (3541, -0.35264754), (4914, -0.31912416), (3285, -0.3041641), (5997, -0.3035813), (5154, -0.2930493), (782, -0.2854037), (1997, -0.28370476), (2902, -0.2817194), (3353, -0.27802846), (414, -0.2773832), (3135, -0.27560827), (3845, -0.27376923), (1455, -0.2735638), (1949, -0.27328032), (4898, -0.26970202), (4435, -0.2694762), (1408, -0.2626838), (1312, -0.26017755), (3106, -0.2592818), (6609, -0.25846046), (7073, -0.2554445), (2927, -0.25359145), (4774, -0.2508814), (5494, -0.25041485), (897, -0.24956687), (5695, -0.24861966), (4274, -0.24600182), (6432, -0.24433468), (3537, -0.2426555), (6488, -0.24260736), (1913, -0.24209917), (2689, -0.24197024), (1779, -0.24076356), (1356, -0.23657046), (4828, -0.23648858), (1867, -0.23616247), (5609, -0.23546898), (1701, -0.23486724), (5006, -0.23440623), (6042, -0.23400354), (4070, -0.23375343), (4590, -0.23365502), (6901, -0.23304757), (5603, -0.2330073), (2453, -0.23271418), (2639, -0.23176712), (2167, -0.23122533), (6566, -

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.04945874214172363 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(782, -0.062149912), (2877, -0.055098794), (5006, -0.052942395), (4435, -0.050154187), (3353, -0.048127912), (6944, -0.047650278), (2267, -0.04720079), (6602, -0.04573367), (4898, -0.045316588), (3457, -0.0446454), (1235, -0.0438995), (1616, -0.043537408), (56, -0.043386523), (4259, -0.042995885), (5247, -0.042953476), (3106, -0.041493684), (1146, -0.041430455), (897, -0.040740117), (2099, -0.04067214), (3785, -0.040503126), (2128, -0.040458925), (2266, -0.040360548), (2062, -0.040354744), (4181, -0.040099725), (4069, -0.040054686), (2915, -0.03932952), (6749, -0.03906725), (6828, -0.039015744), (167, -0.038864106), (1231, -0.038441803), (5494, -0.038142517), (2072, -0.037810676), (4669, -0.037510414), (1849, -0.03750991), (227, -0.03715711), (905, -0.036881354), (6113, -0.036869954), (6785, -0.03643553), (1949, -0.036177352), (1831, -0.03609743), (1350, -0.035935007), (826, -0.035427578), (1325, -0.035333313), (2902, -0.035288185), (6910, -0.035062905), (5904, -0.0347737), (151, -0.0

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.04965496063232422 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(897, -0.25267065), (167, -0.24493036), (3353, -0.21991795), (6910, -0.2179319), (4435, -0.19565237), (4898, -0.19211143), (1949, -0.18512876), (4980, -0.18331344), (782, -0.18259023), (2877, -0.18179694), (5006, -0.1800603), (287, -0.17825276), (2099, -0.17820118), (5904, -0.1779825), (2381, -0.17430776), (826, -0.17215636), (750, -0.16763906), (6707, -0.16445033), (1325, -0.16309804), (6368, -0.1629416), (1235, -0.16218269), (6068, -0.16186239), (3681, -0.15922143), (2745, -0.15847923), (374, -0.1578342), (1408, -0.15703623), (6432, -0.1565528), (3457, -0.15535721), (6519, -0.15211569), (4069, -0.14957818), (5494, -0.1478421), (1831, -0.14630803), (6408, -0.14409435), (2536, -0.14156048), (740, -0.14022997), (3418, -0.13891055), (5997, -0.13844468), (4917, -0.13621786), (1231, -0.13619563), (5264, -0.13488111), (6730, -0.13278237), (6900, -0.13118663), (2915, -0.1303317), (1997, -0.12985353), (5929, -0.12953152), (1025, -0.12931919), (3106, -0.12735665), (6828, -0.12708701), (227, -

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.050633907318115234 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(6910, -0.41330808), (897, -0.393028), (167, -0.37346435), (5904, -0.3709184), (4898, -0.3637581), (7073, -0.35261467), (3127, -0.3512841), (2381, -0.33875906), (3353, -0.33705932), (2902, -0.33513343), (2745, -0.32728928), (2877, -0.3242483), (4435, -0.31944925), (6068, -0.31781664), (826, -0.31747636), (5997, -0.31513768), (1408, -0.3118886), (287, -0.31128013), (2639, -0.31083772), (3106, -0.31019747), (6519, -0.31007883), (1325, -0.3084542), (374, -0.3072728), (4917, -0.3030411), (1956, -0.30270272), (1949, -0.30173525), (5494, -0.30064067), (1235, -0.29886052), (3541, -0.29801288), (3743, -0.2956145), (6368, -0.29309416), (5695, -0.29272786), (2176, -0.29061332), (6951, -0.28902143), (4590, -0.28736487), (4123, -0.2828095), (782, -0.28143376), (3457, -0.27970797), (6828, -0.27778757), (7113, -0.2767111), (6785, -0.27572328), (4219, -0.27142823), (6432, -0.2695613), (5264, -0.26910415), (3681, -0.2678417), (3845, -0.2667903), (6566, -0.26644075), (2462, -0.26531988), (6663, -0.264

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.050167083740234375 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(897, -0.07007375), (782, -0.063665025), (3457, -0.061986264), (3353, -0.061481353), (2099, -0.060956694), (826, -0.060523644), (1231, -0.059063487), (4435, -0.05864393), (167, -0.056661714), (2877, -0.05389099), (4181, -0.053711012), (2381, -0.05348331), (6602, -0.053447843), (4898, -0.053165168), (5904, -0.05224922), (4069, -0.0508513), (2266, -0.05034417), (6944, -0.04997104), (6910, -0.04969901), (1997, -0.049043186), (3681, -0.048815593), (4980, -0.048750795), (2062, -0.04873989), (4212, -0.047233984), (740, -0.046606362), (5783, -0.046405923), (5298, -0.045356747), (4259, -0.044850517), (287, -0.044662952), (5171, -0.04463193), (1068, -0.044543296), (1831, -0.04394509), (2902, -0.043932244), (6408, -0.043472365), (3106, -0.043211475), (1325, -0.043190733), (2882, -0.04302505), (3052, -0.04301074), (750, -0.042813577), (5997, -0.042621363), (6432, -0.04237242), (2443, -0.04157823), (6519, -0.041446347), (7053, -0.041144278), (2406, -0.04113089), (5006, -0.040954217), (1398, -0.04

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.04889512062072754 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(897, -0.19591162), (3353, -0.19189708), (6944, -0.1917024), (6910, -0.18970206), (4629, -0.18562403), (1997, -0.18369797), (3457, -0.18365867), (782, -0.18223892), (4914, -0.17458482), (1949, -0.17439231), (6609, -0.17401326), (2381, -0.17228924), (5997, -0.16651988), (4435, -0.16644561), (5904, -0.1634978), (287, -0.16313806), (4898, -0.16193989), (4070, -0.15871996), (1779, -0.1573568), (2536, -0.1567493), (6042, -0.15499257), (2902, -0.15465578), (3418, -0.15417416), (6793, -0.15223077), (5494, -0.1511484), (6368, -0.14941083), (3681, -0.14920463), (3566, -0.14737886), (4774, -0.14706555), (826, -0.1468907), (2877, -0.14561258), (1312, -0.14456424), (980, -0.14422023), (5154, -0.14286901), (6566, -0.13943644), (1408, -0.13759896), (167, -0.137224), (6113, -0.13645184), (6602, -0.13568519), (3285, -0.13484417), (6965, -0.13416438), (3845, -0.13410044), (1956, -0.13384528), (1235, -0.13366385), (3135, -0.13185881), (4089, -0.13136494), (1325, -0.1308131), (3092, -0.12971212), (1838,

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.04880499839782715 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(167, -0.3925228), (897, -0.37883183), (3353, -0.35808104), (5904, -0.34442848), (6910, -0.34207672), (826, -0.34171468), (6432, -0.34160346), (5006, -0.3409297), (4435, -0.32544157), (1235, -0.31866044), (5609, -0.30650157), (2745, -0.29112232), (5494, -0.29020905), (4980, -0.29009104), (3127, -0.29008096), (3541, -0.2887665), (6068, -0.288615), (753, -0.2881292), (7053, -0.28785163), (6368, -0.2846249), (2877, -0.28100657), (1949, -0.2798398), (3677, -0.27730024), (374, -0.2772993), (4069, -0.27362236), (3566, -0.2734067), (4898, -0.27209347), (2902, -0.27154374), (2381, -0.26691115), (7073, -0.26391947), (1325, -0.2630788), (1867, -0.25972754), (3681, -0.25691444), (1408, -0.25565928), (5264, -0.254471), (782, -0.25440398), (1956, -0.25393134), (4917, -0.2536901), (4332, -0.25270578), (1455, -0.24985966), (3213, -0.24878743), (3457, -0.24795103), (4274, -0.24776654), (6707, -0.24761698), (2536, -0.2468703), (287, -0.24677998), (6519, -0.24659592), (2915, -0.24535649), (5997, -0.244

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.048631906509399414 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(4898, -0.10183306), (2072, -0.082713164), (782, -0.08201208), (2069, -0.07785347), (826, -0.07734004), (1350, -0.07683634), (6944, -0.07544896), (4070, -0.07542921), (5858, -0.07523477), (287, -0.07492328), (449, -0.0738682), (6609, -0.0724168), (2902, -0.07172541), (2957, -0.07145414), (3457, -0.06704006), (551, -0.06688539), (4435, -0.06611844), (5997, -0.065326884), (392, -0.06507917), (6112, -0.06471997), (5672, -0.064383656), (2536, -0.0638143), (3730, -0.063303694), (3353, -0.06288234), (4960, -0.06215838), (5705, -0.061821464), (6722, -0.06069048), (6910, -0.06019821), (4774, -0.05994276), (1956, -0.059759725), (4330, -0.05947431), (3037, -0.059282348), (6602, -0.0591747), (1843, -0.05891054), (6965, -0.05850548), (6068, -0.058438566), (5489, -0.058323886), (3090, -0.05831759), (1997, -0.05802909), (5319, -0.057111457), (3702, -0.05684433), (5546, -0.056283478), (3147, -0.055811904), (2363, -0.055109784), (1867, -0.054910418), (3064, -0.054753464), (6502, -0.05470947), (6423, 

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.05933713912963867 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(5705, -0.32548946), (5997, -0.2870202), (551, -0.27861005), (5284, -0.27632907), (2902, -0.25851718), (6910, -0.25320184), (6944, -0.24849418), (2072, -0.2457405), (1779, -0.24327485), (5095, -0.23676941), (6376, -0.23614125), (2381, -0.23467402), (3353, -0.23277527), (2406, -0.23202375), (1312, -0.22942673), (5858, -0.22855751), (6609, -0.2281654), (2317, -0.2270022), (5154, -0.22672994), (60, -0.22252399), (449, -0.22168837), (3090, -0.22085556), (4898, -0.22072975), (1350, -0.2200652), (6566, -0.2182642), (5263, -0.21813262), (4070, -0.21777159), (4774, -0.21564758), (4738, -0.21443877), (5389, -0.21298343), (5756, -0.20716368), (3457, -0.20397992), (2036, -0.20321025), (2924, -0.20303054), (6828, -0.20285092), (1235, -0.2025537), (5489, -0.2011925), (1125, -0.1990802), (4914, -0.19869485), (4330, -0.19866434), (4808, -0.19779408), (3433, -0.19696744), (2069, -0.1965335), (4435, -0.19606002), (2434, -0.19595031), (3470, -0.19515204), (811, -0.19511409), (5319, -0.19279672), (2266,

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.04932117462158203 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(551, -0.6282501), (6068, -0.54250515), (5937, -0.523625), (6910, -0.5048113), (2317, -0.50309455), (6566, -0.50092363), (6944, -0.48503467), (5756, -0.48415774), (926, -0.48216286), (4994, -0.4733954), (5705, -0.47221053), (4213, -0.46473455), (5997, -0.45909977), (473, -0.45044568), (1312, -0.44758913), (3009, -0.44575366), (6609, -0.44535515), (6113, -0.4419252), (5035, -0.43866304), (2924, -0.43750474), (3700, -0.4350533), (4805, -0.43419617), (2686, -0.43250543), (6574, -0.42762318), (3233, -0.42617843), (2108, -0.4254178), (5977, -0.42015728), (1721, -0.4200754), (6148, -0.41882348), (6397, -0.41707116), (6154, -0.41706938), (3106, -0.41572893), (3845, -0.41071767), (6070, -0.40998444), (1607, -0.409097), (5538, -0.40614477), (6432, -0.40300944), (5465, -0.39546245), (2957, -0.39357004), (1956, -0.39159545), (5049, -0.39097244), (6477, -0.3891764), (170, -0.3885199), (5439, -0.38659084), (3989, -0.383453), (2876, -0.38299578), (1248, -0.38294637), (3736, -0.3758803), (5006, -0.3

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.049642086029052734 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(4898, -0.092276305), (782, -0.08177184), (2072, -0.076811604), (6609, -0.07465867), (1350, -0.07369878), (5997, -0.06990626), (2069, -0.069855124), (5672, -0.06777007), (449, -0.06726655), (551, -0.06588098), (3702, -0.06587741), (287, -0.0657722), (4774, -0.06569245), (4070, -0.06547463), (6602, -0.06527285), (826, -0.06512982), (1843, -0.063869394), (6112, -0.0632449), (6944, -0.063151106), (5156, -0.06053568), (2902, -0.05892344), (6951, -0.05884506), (3353, -0.058646273), (5858, -0.058501884), (5064, -0.058083966), (392, -0.058042303), (2957, -0.057677697), (5489, -0.057395834), (4906, -0.05731405), (3204, -0.057183266), (4435, -0.057072423), (5349, -0.056508597), (2381, -0.056465037), (3457, -0.056013916), (6722, -0.056000803), (2363, -0.05586825), (6965, -0.05485684), (1997, -0.054837283), (1867, -0.054621298), (3147, -0.054354284), (5705, -0.053566527), (5473, -0.053428307), (4219, -0.053405963), (6519, -0.0530957), (3090, -0.05296105), (6641, -0.052766737), (1949, -0.05255936

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.049481868743896484 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(60, -0.35912836), (4898, -0.35332316), (2381, -0.2975124), (2069, -0.2959258), (551, -0.2839195), (1779, -0.2833344), (5154, -0.28268903), (5389, -0.28218994), (3845, -0.27190173), (4774, -0.2705045), (5489, -0.26993272), (6944, -0.26966986), (6376, -0.26777315), (2924, -0.26539811), (2266, -0.26313606), (3090, -0.26275328), (3353, -0.25815794), (750, -0.25611433), (3963, -0.25537232), (7148, -0.2549402), (5705, -0.25444797), (6566, -0.25250864), (2072, -0.25046146), (5095, -0.2501959), (1235, -0.24945214), (4983, -0.24941519), (6951, -0.24853286), (826, -0.24691343), (4994, -0.24576688), (4906, -0.24415927), (1956, -0.243291), (5713, -0.24133317), (2902, -0.23898257), (6722, -0.2379846), (673, -0.2366279), (4332, -0.22817798), (1217, -0.22553286), (2571, -0.2245142), (5997, -0.22414924), (5893, -0.22332361), (3566, -0.22169594), (2786, -0.22097415), (1773, -0.22058435), (2835, -0.21908544), (628, -0.21846537), (5759, -0.21814933), (5672, -0.2174343), (3457, -0.2142473), (1002, -0.21

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.049410104751586914 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(4898, -0.63663137), (551, -0.6314735), (6910, -0.5898627), (5997, -0.55101687), (4808, -0.47203082), (2273, -0.45421198), (750, -0.4539479), (4845, -0.44824412), (5465, -0.44773385), (1217, -0.4426909), (2368, -0.44227812), (2108, -0.44017807), (1721, -0.4356932), (1779, -0.4343061), (7085, -0.43390206), (2406, -0.42950937), (2571, -0.4250625), (628, -0.4241542), (5154, -0.4232416), (59, -0.4218504), (5745, -0.41577548), (2036, -0.4151052), (7044, -0.4147593), (3583, -0.4147304), (3716, -0.41358772), (4942, -0.4125992), (134, -0.41003433), (2069, -0.40962458), (60, -0.4079705), (5033, -0.40627158), (4332, -0.40499166), (4736, -0.40445116), (167, -0.40338373), (3623, -0.40251595), (3470, -0.40139827), (1997, -0.39863598), (3353, -0.39177966), (6068, -0.39025393), (2957, -0.38965622), (6535, -0.38863477), (4416, -0.388338), (6566, -0.3858358), (897, -0.38570765), (1604, -0.38322845), (6220, -0.38223067), (5820, -0.38026267), (7050, -0.38008383), (3090, -0.3783767), (1235, -0.3757194), 

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.04999375343322754 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(4898, -0.08708838), (3457, -0.07909732), (826, -0.076656), (2069, -0.076070376), (782, -0.07603258), (2536, -0.07577176), (6609, -0.07464499), (2072, -0.07462049), (287, -0.07365817), (5858, -0.07232352), (5489, -0.07134039), (6910, -0.07096116), (4774, -0.06999711), (6524, -0.06991905), (6951, -0.06951295), (6944, -0.06800799), (3599, -0.06743139), (3702, -0.0659042), (2957, -0.06517554), (1350, -0.06514202), (3353, -0.065122835), (6112, -0.06286694), (449, -0.06277675), (5156, -0.062419143), (4435, -0.061497815), (6828, -0.061054606), (6519, -0.06064457), (2902, -0.060054395), (4960, -0.059954595), (3064, -0.059870042), (3037, -0.059803814), (5997, -0.059641737), (897, -0.059216775), (2381, -0.059157412), (5672, -0.05891623), (6304, -0.058567774), (6804, -0.05845882), (3937, -0.0584029), (3831, -0.058373198), (1843, -0.05793701), (3090, -0.05788108), (2363, -0.057730608), (6722, -0.057658896), (551, -0.057614867), (1956, -0.057123736), (392, -0.056734115), (5820, -0.05648756), (60,

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.053266048431396484 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(5489, -0.32786804), (5705, -0.3061462), (1312, -0.2969622), (5389, -0.29123163), (6951, -0.28623584), (551, -0.27224612), (2072, -0.2639602), (5154, -0.25620613), (1779, -0.25366172), (6910, -0.24954869), (4898, -0.2371146), (5997, -0.2292682), (1235, -0.22902206), (2877, -0.22772281), (60, -0.2271719), (4435, -0.22611815), (2069, -0.22460358), (5284, -0.22421981), (1350, -0.22009559), (6376, -0.21800582), (4774, -0.21671602), (2381, -0.21597077), (2957, -0.21555378), (3845, -0.21307576), (811, -0.2123119), (2902, -0.21169463), (1169, -0.21131329), (6295, -0.20748444), (2686, -0.20509762), (1231, -0.20470242), (5820, -0.2015666), (4808, -0.19754972), (3356, -0.19651845), (5156, -0.1964361), (3353, -0.19616172), (5319, -0.19600089), (6944, -0.19521338), (6609, -0.19432567), (2266, -0.1940755), (1256, -0.19131182), (6566, -0.19101426), (194, -0.19091853), (2571, -0.19044678), (3728, -0.18876763), (4906, -0.18801916), (3090, -0.18789397), (3583, -0.18381663), (1768, -0.18375118), (1596,

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.05071902275085449 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(1779, -0.5428279), (2957, -0.5019162), (2902, -0.48973882), (4898, -0.48656803), (3353, -0.47543252), (5986, -0.47018936), (5858, -0.46950883), (2877, -0.46258262), (1312, -0.45874146), (5705, -0.4503346), (1231, -0.449513), (465, -0.44945964), (4906, -0.44806743), (4137, -0.44686735), (1848, -0.44492006), (2381, -0.4371832), (5904, -0.419756), (1997, -0.41804406), (3293, -0.4137541), (6566, -0.41206867), (3681, -0.41046697), (551, -0.4104659), (5997, -0.40485683), (3106, -0.40253237), (6951, -0.39779854), (7053, -0.39583582), (4274, -0.3949146), (750, -0.39254808), (1589, -0.39164472), (5518, -0.388472), (3979, -0.38501176), (4435, -0.3842562), (3285, -0.3817111), (5489, -0.37854132), (2835, -0.3775026), (6519, -0.37605417), (6068, -0.37546265), (3457, -0.375018), (414, -0.37293255), (6593, -0.37289196), (7050, -0.3710128), (4808, -0.367865), (3583, -0.3654284), (1217, -0.36525238), (4330, -0.36393154), (6295, -0.3595282), (2273, -0.35881507), (4784, -0.35718575), (5756, -0.3561962)

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.049041748046875 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(4898, -0.13956258), (2069, -0.11791932), (3937, -0.11062654), (60, -0.11009408), (3470, -0.107370995), (6148, -0.10649894), (750, -0.105733685), (6804, -0.10363156), (5389, -0.10157136), (6163, -0.10094741), (1681, -0.100253895), (3009, -0.10011268), (5154, -0.09739412), (1768, -0.09721236), (3595, -0.09687917), (4815, -0.096804775), (5284, -0.09347758), (1235, -0.09344071), (1231, -0.093076944), (57, -0.09218821), (2266, -0.09188707), (59, -0.09109203), (5527, -0.090709716), (1589, -0.08906241), (6151, -0.08747252), (250, -0.08636898), (4207, -0.08611634), (6687, -0.08593825), (4181, -0.085276954), (5319, -0.084120214), (4805, -0.08390748), (2416, -0.08382799), (2381, -0.0835056), (1327, -0.08174124), (2896, -0.080990985), (5858, -0.08080568), (1699, -0.08030542), (5263, -0.08024624), (7148, -0.080105714), (1362, -0.078901395), (3716, -0.07853675), (4942, -0.07828439), (4084, -0.07812297), (903, -0.07759302), (3599, -0.07759139), (3442, -0.07686795), (4774, -0.07633801), (1345, -0.0

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.05078911781311035 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(2069, -0.39183924), (1327, -0.3523024), (6092, -0.34500337), (7013, -0.34491268), (4898, -0.34283215), (320, -0.33900777), (1169, -0.33011594), (1281, -0.32500017), (4774, -0.3162563), (6519, -0.31591684), (5263, -0.3157772), (4815, -0.31520516), (6816, -0.31505582), (750, -0.31487957), (6268, -0.31006366), (7148, -0.3093139), (3875, -0.30899316), (4443, -0.3073376), (5154, -0.30654317), (5936, -0.30174032), (2171, -0.2958561), (6743, -0.29275942), (5820, -0.2915425), (2317, -0.2899469), (3470, -0.28829008), (16, -0.28554067), (6113, -0.2843733), (4207, -0.28404164), (1231, -0.2839611), (3599, -0.28105408), (4880, -0.28005135), (46, -0.27989644), (5705, -0.27983406), (2413, -0.2783099), (5503, -0.27777278), (59, -0.27599624), (6037, -0.27476045), (4906, -0.2730248), (1768, -0.27135104), (3457, -0.27112633), (4567, -0.2663067), (5284, -0.26597917), (3504, -0.2657373), (2205, -0.2652712), (5937, -0.26313886), (3989, -0.26312467), (3601, -0.2618016), (6030, -0.26151484), (3882, -0.26012

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.05377316474914551 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(7044, -0.6025393), (5936, -0.5846611), (4994, -0.58353907), (164, -0.5819019), (2368, -0.5636939), (5263, -0.55347335), (6548, -0.5528495), (3344, -0.5276266), (7156, -0.5180859), (1681, -0.51485413), (3504, -0.5101111), (1390, -0.5026856), (4235, -0.50255424), (7142, -0.4965853), (2924, -0.4965548), (5705, -0.48687083), (2166, -0.47495085), (4766, -0.4742179), (2176, -0.4721858), (3989, -0.471328), (397, -0.46829587), (406, -0.4671909), (5926, -0.4659738), (59, -0.46495882), (1169, -0.4637898), (6220, -0.46233538), (3353, -0.4618405), (2549, -0.46105164), (1054, -0.45834824), (1217, -0.45740977), (2963, -0.45537233), (385, -0.45389163), (1768, -0.45136058), (375, -0.44895303), (727, -0.44467527), (210, -0.4376139), (6566, -0.43043593), (134, -0.42940828), (3730, -0.42932367), (5035, -0.42905545), (2671, -0.4261676), (6754, -0.42607594), (6148, -0.41876328), (5971, -0.4181832), (3845, -0.41670376), (3450, -0.4125651), (2205, -0.4100261), (3728, -0.40735817), (2381, -0.40663266), (194

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.06705594062805176 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(4898, -0.12614979), (2069, -0.11406091), (3442, -0.10718529), (1681, -0.105143666), (3937, -0.10278821), (1327, -0.100978404), (3470, -0.10091412), (5820, -0.097216226), (5705, -0.09656499), (59, -0.09647906), (3107, -0.095108524), (6163, -0.09377375), (897, -0.09370941), (6148, -0.09077971), (60, -0.090257406), (5284, -0.08916578), (3037, -0.08900137), (2038, -0.0882224), (5319, -0.08796544), (6871, -0.08782974), (1726, -0.086865366), (6964, -0.08683222), (1169, -0.086656414), (2930, -0.08607131), (1997, -0.08594644), (6376, -0.08529766), (5154, -0.083341524), (2099, -0.081322335), (4181, -0.08130516), (309, -0.080524534), (680, -0.08034742), (3127, -0.07966584), (1768, -0.07959341), (1312, -0.07920415), (2381, -0.07880761), (57, -0.07785963), (5858, -0.07760008), (1111, -0.07756399), (1390, -0.077563405), (750, -0.07735791), (4815, -0.077186175), (5756, -0.077011295), (4444, -0.07687641), (3353, -0.07664046), (4942, -0.07655575), (5527, -0.07584668), (4435, -0.07555735), (2924, -0.

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.05055737495422363 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(2317, -0.39236733), (2069, -0.36586314), (60, -0.3645083), (3875, -0.36321938), (2108, -0.35794187), (1327, -0.34439188), (6376, -0.33626097), (4898, -0.33138594), (5756, -0.31819904), (4622, -0.3179111), (1054, -0.31555086), (1768, -0.31494522), (2167, -0.29915968), (3090, -0.29242194), (6743, -0.29115036), (3066, -0.29027316), (750, -0.28954637), (1681, -0.28913683), (5039, -0.2864578), (1144, -0.2825989), (2178, -0.28230152), (5489, -0.28193825), (5033, -0.27814072), (3752, -0.27669775), (2924, -0.276134), (1235, -0.27517247), (4207, -0.27453417), (6316, -0.272318), (4320, -0.26955616), (1773, -0.26775053), (6682, -0.2660751), (1169, -0.26523677), (304, -0.2646146), (767, -0.2641952), (5893, -0.26374376), (6964, -0.2631876), (6186, -0.26310706), (4119, -0.26232672), (5926, -0.26000375), (7013, -0.25998655), (5424, -0.25865525), (5154, -0.25799495), (1997, -0.257492), (2205, -0.25534463), (2036, -0.2534035), (826, -0.2525445), (5527, -0.2517811), (6126, -0.25157547), (2416, -0.2498

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.05131387710571289 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(750, -0.70215744), (60, -0.65629596), (2930, -0.60063905), (2542, -0.54962194), (5154, -0.5387048), (5295, -0.52552176), (1201, -0.5203767), (5218, -0.5192243), (5389, -0.51309973), (4119, -0.51024765), (2069, -0.49660966), (2317, -0.49365833), (6316, -0.48971903), (2368, -0.4894349), (2164, -0.4866433), (368, -0.4784364), (6113, -0.47618774), (6944, -0.47392792), (5641, -0.46350884), (1178, -0.46240273), (829, -0.46188432), (5756, -0.45379254), (2653, -0.45258301), (4898, -0.44312978), (4575, -0.43640906), (6712, -0.43347615), (6574, -0.43206885), (6951, -0.42861795), (5049, -0.42777318), (4137, -0.4238898), (3924, -0.41883606), (4070, -0.4176903), (241, -0.41751203), (4866, -0.41091776), (5858, -0.41023323), (92, -0.4062779), (2796, -0.40538633), (3631, -0.40238366), (5986, -0.40124646), (4331, -0.40116182), (6790, -0.4011074), (4980, -0.4003101), (5629, -0.3994431), (5473, -0.39814815), (5221, -0.3973197), (6485, -0.3969678), (277, -0.3934934), (3470, -0.39294454), (3064, -0.39126

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.05150198936462402 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(4898, -0.13898098), (6163, -0.11249457), (3937, -0.11184898), (5820, -0.11175892), (2069, -0.10845354), (1327, -0.10359795), (57, -0.103009306), (1235, -0.100055836), (1231, -0.09923573), (3599, -0.098513275), (3470, -0.09585824), (1169, -0.09581494), (449, -0.09478751), (1681, -0.09388415), (4181, -0.09034598), (6687, -0.090262115), (59, -0.09016195), (60, -0.08944245), (1726, -0.08881848), (1768, -0.0887389), (6551, -0.08757396), (5858, -0.08733076), (5154, -0.086845964), (2930, -0.08669237), (1362, -0.08639935), (680, -0.08618832), (4774, -0.08583051), (5389, -0.08581294), (5705, -0.0847325), (6151, -0.083972335), (3090, -0.08338964), (3215, -0.083331265), (6964, -0.08267125), (860, -0.08266668), (6376, -0.082165495), (4084, -0.082087144), (1763, -0.081819415), (5319, -0.08181514), (5489, -0.08178363), (6148, -0.08165842), (5284, -0.08066521), (3442, -0.08058318), (5675, -0.08041663), (750, -0.08035125), (3716, -0.08016231), (4942, -0.07941596), (5936, -0.07919413), (2892, -0.0790

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.04845714569091797 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(4898, -0.44524014), (2069, -0.4060261), (6871, -0.36362535), (5284, -0.3438486), (6154, -0.33183503), (1235, -0.32393873), (6548, -0.31789577), (826, -0.3131903), (2925, -0.30954835), (6220, -0.30850115), (7013, -0.30509838), (1768, -0.30402285), (5893, -0.30060303), (60, -0.2997443), (59, -0.28919974), (4994, -0.2882564), (304, -0.284407), (1913, -0.28302848), (5035, -0.2817749), (7148, -0.28066757), (3344, -0.28011698), (2317, -0.27956697), (6591, -0.27939147), (6068, -0.27790046), (5641, -0.27645966), (1312, -0.27499795), (3404, -0.27110624), (5629, -0.2710791), (2038, -0.27107707), (1997, -0.26935577), (750, -0.26871765), (2574, -0.2672596), (7053, -0.2669547), (6293, -0.2656713), (1532, -0.26293272), (1202, -0.26153028), (6566, -0.25911492), (5319, -0.25850597), (1589, -0.25801724), (4774, -0.25708443), (6535, -0.25572085), (1169, -0.25412175), (4815, -0.2531256), (3377, -0.25108108), (2416, -0.24938722), (4991, -0.24879752), (6027, -0.24629961), (3963, -0.24604511), (2381, -0.2

/opt/homebrew/Caskroom/miniconda/base/lib/python3.12/site-packages/implicit/utils.py:164: ParameterWarning: Method expects CSR input, and was passed csc_matrix instead. Converting to CSR took 0.04874110221862793 seconds
  warnings.warn(


  0%|          | 0/15 [00:00<?, ?it/s]

[(4898, -0.6015213), (2925, -0.5640896), (3755, -0.5530015), (2924, -0.518721), (4942, -0.508617), (4991, -0.49120906), (1281, -0.48934618), (716, -0.48870647), (6951, -0.47688976), (4994, -0.4724596), (1629, -0.46617883), (6793, -0.46562603), (5705, -0.46290866), (1447, -0.4557876), (1064, -0.4526959), (2106, -0.45129174), (2693, -0.44904768), (5489, -0.447216), (2835, -0.44700035), (6566, -0.44608754), (2381, -0.44520396), (5756, -0.44452083), (1779, -0.4431992), (5284, -0.4412259), (4435, -0.43424818), (5413, -0.43277535), (626, -0.4308325), (1768, -0.4217257), (6048, -0.42119437), (4736, -0.41972998), (5423, -0.41920573), (4452, -0.41187137), (6904, -0.4110135), (795, -0.41054833), (1106, -0.40867117), (6113, -0.40696573), (6376, -0.4041444), (1178, -0.39760137), (2266, -0.3954322), (5631, -0.39412618), (1656, -0.39353287), (1203, -0.39263308), (2069, -0.39165547), (3521, -0.39136022), (2687, -0.39112014), (1220, -0.39005172), (390, -0.38808596), (3146, -0.3877945), (6910, -0.38699

You can now [go back to the main notebook](../FinalProject_Notebook.ipynb).